File: greedyAlgorithms/classroomScheduling.py
"""
Grokking Algorithms Ch 8: Greedy Algorithms
The classroom scheduling problem
"""

classes = [
  {
    'name': 'ART',
    'start': '9:00',
    'end': '10:00'
  },
  {
    'name': 'ENG',
    'start': '9:30',
    'end': '10:30'
  },
  {
    'name': 'MATH',
    'start': '10:00',
    'end': '11:00'
  },
  {
    'name': 'CS',
    'start': '10:30',
    'end': '11:30'
  },
  {
    'name': 'MUSIC',
    'start': '11:00',
    'end': '12:00'
  },
]

def compare_times(old_time, new_time):
  old_time_end = old_time['end'].split(':')
  new_time_start = new_time['start'].split(':')

  if (int(old_time_end[0]) < int(new_time_start[0])):
    return True
  if (int(old_time_end[0]) == int(new_time_start[0])):
    if (int(old_time_end[1]) <= int(new_time_start[1])):
      return True
  return False

def findBestSchedule(class_list):
  ## negation and data type checking
  if not isinstance(class_list, list):
    return 'Please pass in a list!!'

  schedule = []
  latest_class = None

  for current_class in class_list:
    if len(schedule) == 0:
      schedule.append(current_class['name'])
      latest_class = current_class
    if latest_class['name'] != current_class['name'] and compare_times(latest_class, current_class):
      schedule.append(current_class['name'])
      latest_class = current_class
  return schedule

print(findBestSchedule(classes)) ## should return [ART, MATH, MUSIC]
print(findBestSchedule({'Art': '10:00', 'Gym': '112'})); ## should return 'Please pass in an array!!'

File: greedyAlgorithms/classroomScheduling.js
/*
Grokking Algorithms Ch 8: Greedy Algorithms
The classroom scheduling problem
*/

const classes = [
  {
    'name': 'ART',
    'start': '9:00',
    'end': '10:00'
  },
  {
    'name': 'ENG',
    'start': '9:30',
    'end': '10:30'
  },
  {
    'name': 'MATH',
    'start': '10:00',
    'end': '11:00'
  },
  {
    'name': 'CS',
    'start': '10:30',
    'end': '11:30'
  },
  {
    'name': 'MUSIC',
    'start': '11:00',
    'end': '12:00'
  },
];

var compareTimes = (oldTime, newTime) => {
  // if newTime's start is >= oldTime's end
    // return true
  let oldTimeEnd = oldTime.end.split(':');
  let newTimeStart = newTime.start.split(':')
  if (Number(oldTimeEnd[0]) < Number(newTimeStart[0])) {
    return true;
  };
  if (Number(oldTimeEnd[0]) === Number(newTimeStart[0])) {
    if (Number(oldTimeEnd[1]) <= Number(newTimeStart[1])) {
      return true;
    };
  };
  return false;
};

var findBestSchedule = (classList) => {
  if (!Array.isArray(classList)) {
    return 'Please pass in an array!!';
  };
  let schedule = [];
  var latestClass;
  // for-in loop on an array
  for (var currentClass in classList) {
    if (schedule.length === 0) {
      schedule.push(classList[currentClass].name);
      latestClass = currentClass;
    }
    if (latestClass !== currentClass && compareTimes(classList[latestClass], classList[currentClass])) {
      schedule.push(classList[currentClass].name);
      latestClass = currentClass;
    }
  };
  return schedule;
};

console.log(findBestSchedule(classes)); //should return [ART, MATH, MUSIC]
console.log(findBestSchedule({'Art': '10:00', 'Gym': '112'})); //should return 'Please pass in an array!!'